### PR TITLE
refactor: Basilisk hs integration

### PR DESCRIPTION
**Pull request by manoj-juspay**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

## Description
Legacy locker will be depreciated as it doesn't have the key rotation feature,
We will be using basilisk-hs card vault to store card and other payment method

### Additional Changes

- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?
Tested it locally with basiliks-hs setup , we need to test basilisk-v3 as this is the services effected with the changes
Lets do it in custom pods post review

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payment_methods/cards.rs [hunk 1]
```diff
@@ -125,73 +135,156 @@ pub async fn update_customer_payment_method(
...
     Ok(payment_method_resp)
 }
 
+#[instrument(skip_all)]
+pub async fn get_card_from_hs_locker<'a>(
+    state: &'a routes::AppState,
+    customer_id: &str,
+    merchant_id: &str,
+    card_reference: &'a str,
+) -> errors::RouterResult<payment_methods::Card> {
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/cards.rs:218]: Can we have the error, be something other than, `ApiErrorResponse`, as having errors like, VaultError, ApiClientError, and ParsingError in the current context. I suggest using some other error variant while returning as we can later convert it to ApiErrorResponse in the parent function

File: crates/router/src/core/payment_methods/transformers.rs [hunk 2]
```diff
@@ -1,104 +1,220 @@
...
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DeleteCardResp {
+    pub status: String,
+    pub error_message: Option<String>,
+    pub error_code: Option<String>,
+}
+
+pub fn basilisk_hs_key_id() -> &'static str {
+    "1"
+}
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:76]: Why is this hard coded? can we fetch it fron configuration?

File: crates/router/src/core/payment_methods/transformers.rs [hunk 3]
```diff
@@ -1,104 +1,220 @@
...
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Error on jwe encrypt")?;
+    let jwe_payload: Vec<&str> = jwe_encrypted.split('.').collect();
+    Ok(encryption::JweBody {
+        header: jwe_payload[0].to_owned(),
+        iv: jwe_payload[2].to_owned(),
+        encrypted_payload: jwe_payload[3].to_owned(),
+        tag: jwe_payload[4].to_owned(),
+        encrypted_key: jwe_payload[1].to_owned(),
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/transformers.rs:159]: If there is any issue, while splitting this, the code could panic. Can we use some safer functionality like `.get()` or have a implementation of `FromStr` something like `impl FromStr for JweBody`?

File: crates/router/src/core/payment_methods/vault.rs [hunk 4]
```diff
@@ -381,6 +382,47 @@ impl Vault {
...
+        &keys.locker_key_identifier1
+    } else {
+        &keys.locker_key_identifier2
+    }
+}
+
+#[cfg(feature = "basilisk")]
+pub async fn get_locker_jwe_keys(
+    keys: &Jwekey,
+) -> CustomResult<(String, String), errors::EncryptionError> {
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/vault.rs:397]: Should be have them as `Secret<(String, String)>` or `StrongSecret<(String, String)>`?

File: crates/router/src/services/encryption.rs [hunk 5]
```diff
@@ -50,6 +50,24 @@ impl NonceSequence for NonceGen {
...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JweBody {
+    pub header: String,
+    pub iv: String,
+    #[serde(rename = "encryptedPayload")]
+    pub encrypted_payload: String,
+    pub tag: String,
+    #[serde(rename = "encryptedKey")]
+    pub encrypted_key: String,
+}
```
Comment (NishantJoshi00) [inline crates/router/src/services/encryption.rs:69]: can we use `#[serde(rename_all = "camelCase")]`?

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/transformers.rs (hunk 2 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/transformers.rs]: hardcoded in basilisk-hs as well , will keep in sync with basilisk

File: crates/router/src/services/encryption.rs (hunk 5 above)
Comment (manoj-juspay) [reply crates/router/src/services/encryption.rs:69]: yes

File: crates/router/src/core/payment_methods/transformers.rs (hunk 3 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/transformers.rs:161]: added check on length

File: crates/router/src/core/payment_methods/vault.rs (hunk 4 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/vault.rs:392]: not required

File: crates/router/src/core/payment_methods/cards.rs [hunk 6]
```diff
@@ -139,74 +149,158 @@ pub async fn update_customer_payment_method(
...
-            Ok(card) => card
+    )
+    .await?;
+    let response = services::call_connector_api(state, request)
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)?;
+    let stored_card_response = match response {
+        Ok(card) => {
+            let jwe_body: services::JweBody = card
                 .response
```
Comment (NishantJoshi00) [inline crates/router/src/core/payment_methods/cards.rs:186]: What happened to `mock_locker`? are we doing this switch somewhere else?

Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/cards.rs]: yes added

File: crates/router/src/core/payment_methods/vault.rs [hunk 7]
```diff
@@ -374,6 +377,47 @@ impl Vault {
...
+            keys.locker_encryption_key1.to_string(),
+        )
+        .await?;
+        let private_key = services::KeyHandler::get_kms_decrypted_key(
+            keys,
+            keys.locker_decryption_key1.to_string(),
+        )
+        .await?;
+        Ok((public_key, private_key))
+    } else {
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/vault.rs:406]: we need to check `locker_key_identifier2` also equal to key_id, if that fails we need to throw proper error and metrics too.

File: crates/router/src/core/payment_methods/vault.rs [hunk 8]
```diff
@@ -374,6 +377,47 @@ impl Vault {
...
+            keys,
+            keys.locker_encryption_key1.to_string(),
+        )
+        .await?;
+        let private_key = services::KeyHandler::get_kms_decrypted_key(
+            keys,
+            keys.locker_decryption_key1.to_string(),
+        )
+        .await?;
+        Ok((public_key, private_key))
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/vault.rs:405]: Make `private_key` as secret or make the function visibility as private.

File: crates/router/src/services.rs [hunk 9]
```diff
@@ -1,6 +1,6 @@
 pub mod api;
 pub mod authentication;
-#[cfg(feature = "basilisk")]
+// #[cfg(feature = "basilisk")]
```
Comment (jarnura) [inline crates/router/src/services.rs:3]: remove commented code, why we are moving out from basilisk feature?

File: crates/router/src/services.rs [hunk 10]
```diff
@@ -9,7 +9,7 @@ use std::sync::{atomic, Arc};
 use redis_interface::{errors::RedisError, PubSubInterface};
 
 pub use self::api::*;
-#[cfg(feature = "basilisk")]
+// #[cfg(feature = "basilisk")]
```
Comment (jarnura) [inline crates/router/src/services.rs:12]: remove commented code, why we are moving out from basilisk feature?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/transformers.rs [hunk 11]
```diff
@@ -1,104 +1,229 @@
...
+pub async fn mk_basilisk_req(
+    jwekey: &Jwekey,
+    jws: &str,
+) -> CustomResult<encryption::JweBody, errors::VaultError> {
+    let jws_payload: Vec<&str> = jws.split('.').collect();
+    when(jws_payload.len().ne(&3), || {
+        Err(errors::VaultError::SaveCardFailed)
+    })?;
+    let jws_body = encryption::JwsBody {
+        header: jws_payload[0].to_owned(),
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/transformers.rs:139]: We should not use direct indexing this cause panic's. Iterate and take values, if not found throw error.

Decoupling using `when(jws_payload.len().ne(&3)` solve that but the code still has a way to panic with this check remove or bypassed in future.

File: crates/router/src/core/payment_methods/transformers.rs [hunk 12]
```diff
@@ -1,104 +1,229 @@
...
+    let jwe_encrypted = encryption::encrypt_jwe(jwekey, &payload, public_key)
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Error on jwe encrypt")?;
+    let jwe_payload: Vec<&str> = jwe_encrypted.split('.').collect();
+    when(jwe_payload.len().ne(&5), || {
+        Err(errors::VaultError::SaveCardFailed)
+    })?;
+    Ok(encryption::JweBody {
+        header: jwe_payload[0].to_owned(),
```
Comment (jarnura) [inline crates/router/src/core/payment_methods/transformers.rs:163]: Avoid using direct indexing.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/vault.rs (hunk 7 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/vault.rs:406]: sure

File: crates/router/src/core/payment_methods/vault.rs (hunk 8 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/vault.rs:405]: making fn private

File: crates/router/src/services.rs (hunk 9 above)
Comment (manoj-juspay) [reply crates/router/src/services.rs:3]: this makes the entire encryption module non usable in local, we have fn like jwe, jws which works the same in every env

File: crates/router/src/services.rs (hunk 10 above)
Comment (manoj-juspay) [reply crates/router/src/services.rs:12]: +1

File: crates/router/src/core/payment_methods/transformers.rs (hunk 11 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/transformers.rs:139]: sure

File: crates/router/src/core/payment_methods/transformers.rs (hunk 12 above)
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/transformers.rs:163]: sure

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/core/payment_methods/transformers.rs [hunk 13]
```diff
@@ -1,104 +1,255 @@
...
+    let tag = jwe.tag;
+    format!("{header}.{encryption_key}.{iv}.{encryption_payload}.{tag}")
+}
+
+pub fn get_dotted_jws(jws: encryption::JwsBody) -> String {
+    let header = jws.header;
+    let payload = jws.payload;
+    let signature = jws.signature;
+    format!("{header}.{payload}.{signature}")
+}
```
[the remark this answers is not in the document]
Comment (manoj-juspay) [reply crates/router/src/core/payment_methods/transformers.rs:95]: lets refactor the code once we have everything merge in main along with nishant legacy locker revert changes

**Review by SanchithHegde — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**